build.sbt

```diff
@@ -27,3 +25,299 @@ libraryDependencies ++= {
+      "-D__int64=\"long long\" " +
+      "-I " + "\"" + jdkIncludePath.absolutePath + "\" " +
+      "-I " + "\"" + jdkWinIncludePath.absolutePath + "\" " +
+      "-I " + "\"" + z3BinFilePath.getParentFile.absolutePath + "\" " +
+      "-Wreturn-type ", log)
+
+  } else if (isMac) {
+    val frameworkPath = "/System/Library/Frameworks/JavaVM.framework/Versions/Current/Headers"
+
+    exec("install_name_tool -id @loader_path/"+z3Name+" "+z3BinFilePath.absolutePath, log)
+    exec("install_name_tool -id @loader_path/"+javaZ3Name+" "+javaZ3BinFilePath.absolutePath, log)
+    // make the dependency to z3 be relative to the caller's location
+    exec("install_name_tool -change "+z3Name+" @loader_path/"+z3Name+" "+javaZ3BinFilePath.absolutePath, log)
+
+    exec("gcc -std=gnu89 -o " + libBinFilePath.absolutePath + " " +
+      "-dynamiclib" + " " +
+      "-install_name @loader_path/"+soName + " " +
+      "-I" + jdkIncludePath.absolutePath + " " +
+      "-I" + jdkMacIncludePath.absolutePath + " " +
+      "-I" + frameworkPath + " " +
+      "-L" + z3BuildPath.absolutePath + " " +
+      "-g -lc " +
+      "-Wl,-rpath,"+extractDir(checksumKey.value)+" " +
+      "-lz3 -fPIC -O2", log)
+
+  } else {
+    sys.error("Unknown arch: "+osInf+" - "+osArch)
+  }
+  () // unit task!
+}
+
+val packageTask = (Keys.`package` in Compile).dependsOn(gccKey)
+
+val newMappingsTask = Def.task {
+  val normalFiles = (mappings in (Compile, packageBin)).value
+  val log = streams.value.log
+
+  val newBinaryFiles = (libBinFilePath +: z3BinaryFiles).map { f => f.getAbsoluteFile -> ("lib-bin" + DS + f.getName) }
+
+  log.info("Bundling binary files:")
+  for ((from, to) <- newBinaryFiles) {
+    log.info(" - " + from + " -> " + to)
+  }
+
+  log.info("Bunding relevant java-Z3 files:")
+  val outputDir = new File(System.getProperty("java.io.tmpdir") + DS + "Z3JAR_jars" + DS)
+  outputDir.delete
+  outputDir.mkdirs
+  IO.unzip(z3JarFile, outputDir)
+
+  val z3JavaDepsMappings: Seq[(File, String)] = listAllFiles(outputDir).flatMap { f =>
+    if (f.isDirectory) None else {
+      import java.nio.file.Paths
+      val path = Paths.get(outputDir.getAbsolutePath).relativize(Paths.get(f.getAbsolutePath)).toString
+      val extensionSplit = path.split("\\.")
+      if (extensionSplit.length < 2 || extensionSplit(1) != "class") None else {
+        val classPath = extensionSplit(0).replace(DS, ".")
+        if (z3JavaDepsPrefixes.exists(prefix => classPath.startsWith(prefix))) {
+          log.info(" - " + classPath)
+          Some(f.getAbsoluteFile -> path)
+        } else {
+          None
+        }
+      }
+    }
+  }
+
+  newBinaryFiles ++ z3JavaDepsMappings ++ normalFiles
+}
+
+lazy val root = (project in file("."))
+    .settings(
+      checksumKey := checksumTask.value,
+      gccKey := gccTask.value,
+      z3Key := z3Task.value,
+      Keys.`package` in Compile := packageTask.value,
+      unmanagedJars in Compile += Attributed.blank(z3JarFile),
```

Comment:
I think this line doesn't work as intended (but it's the same as before): the `z3JarFile` is just a string, and if the Z3 project wasn't compiled yet, `unamangedJars` will contain the right path to `com.microsoft.z3.jar` but the file won't be there, and compilation will simply fail. It would work the next time though, if somehow Z3 was built at least once and the jar is there.

A potential fix might be to make the `z3JarFile` be a task that dependsOn `z3.value`